src/screens/MainApp/DiagnoseResponse/index.js

```diff
@@ -1,12 +1,69 @@
-import React from 'react'
-import { Text, View } from 'react-native'
+import React, { useState, useEffect } from 'react'
+import { View, ScrollView, SafeAreaView, ActivityIndicator } from 'react-native'
+import Diagnose from './Diagnose'
+import * as DatabaseService from '~/databaseService'
+import * as StorageService from '~/storageService'
+import { ForceRerenderOnNavigation } from '~/navigationService'
+import styles from './styles'
+
+const getDiagnosesFromDatabase = async () => {
+  try {
+    return await DatabaseService.getAnsweredDiagnosesForCurrentUser()
+  } catch (error) {
+    return []
+  }
+}
+
+const getURL = async (imageReference) => {
+  try {
+    return await StorageService.getDownloadURL(imageReference)
+  } catch (error) {
+    return null
+  }
+}
 
 const DiagnoseResponse = () => {
+  const [diagnoses, setDiagnoses] = useState(null)
+  const [downloadingDiagnoses, setDownloadingDiagnoses] = useState(true)
+  const [refetchTrigger, toggleRefetch] = useState(false)
+
+  const refetchDiagnoses = () => {
+    // Tab navigator doesn't unmount components, we need this to refetch need data when pressing DiagnoseResponse tab.
+    // alternative is to add a query snapshot listener, so that new changes in the database trigger a re-render automatically.
+    toggleRefetch(!refetchTrigger)
+  }
+
+  const buildDiagnoses = async () => {
+    setDownloadingDiagnoses(true)
+    const answers = await getDiagnosesFromDatabase()
+    const diagnoses = []
+    for (let i = 0; i < answers.length; i++) {
```

Comment:
Consider using Array's methods instead.

Reply:
Is this regarding forEach? I run into a weird problem using async/await inside forEach. Like it didn't execute due to each forEach function being a Promise.

I tried something like:

`answers.forEach(async(answer)=>{...})`

But I never specify that I should wait for the functions inside the forEach to finish. Do you suggest I use something like the following:

```
Promise.all(
answers.forEach(...)
)
```
**Note: I added the screencast too.**

Reply:
Yes, it can be done with Array methods and `Promise.all`, but you need to use `map` instead of `forEach`. Also keep in mind you'll be doing operations in parallel. You _can_ force the async operations to be run in sequence with `reduce`. The current behavior with your `for` loop is to run them in sequence. Example:
```javascript
// in parallel
const diagnoses = await Promise.all(answers.map(async answer => {
  // some async operation to return the diagnose
}))

// in sequence
const diagnoses = await answers.reduce(
  async (diagnoses, answer) => {
    // diagnoses will be a promise that resolves to the result of the previous reduction
    diagnoses = await diagnoses // wait for the previous diagnose to finish
   
    const diagnose = await somehowGetTheDiagnoseAsynchronously() // get the new diagnose

    diagnoses.push(diagnose) // push the diagnose to the diagnoses array

    return diagnoses // return the diagnoses array with the new diagnose
  },
  Promise.resolve([])
)
```